yrs/src/doc.rs

```diff
@@ -295,58 +296,79 @@ pub enum OffsetKind {
-    fn try_transact(&self) -> Result<Transaction, BorrowError>;
+    fn try_transact(&self) -> Result<Transaction, TransactionAcqError>;
 
     /// Creates a transaction used for all kind of block store operations.
     /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
-    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError>;
+    fn try_transact_mut(&self) -> Result<TransactionMut, TransactionAcqError>;
 
     /// Creates a transaction used for all kind of block store operations.
     /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
     fn transact(&self) -> Transaction {
-        self.try_transact().expect("cannot read document store contents because another read-write transaction is in progress")
+        self.try_transact().unwrap()
     }
 
     /// Creates a transaction used for all kind of block store operations.
     /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
     fn transact_mut(&self) -> TransactionMut {
-        self.try_transact_mut()
-            .expect("only one read-write transaction can be active at the same time")
+        self.try_transact_mut().unwrap()
     }
 }
 
 impl Transact for Doc {
-    fn try_transact(&self) -> Result<Transaction, BorrowError> {
+    fn try_transact(&self) -> Result<Transaction, TransactionAcqError> {
         Ok(Transaction::new(self.store.try_borrow()?))
     }
 
-    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError> {
+    fn try_transact_mut(&self) -> Result<TransactionMut, TransactionAcqError> {
         Ok(TransactionMut::new(self.store.try_borrow_mut()?))
     }
 }
 
 impl Transact for Branch {
-    fn try_transact(&self) -> Result<Transaction, BorrowError> {
+    fn try_transact<'a>(&'a self) -> Result<Transaction<'a>, TransactionAcqError> {
         let store = self.store.as_ref().unwrap();
-        Ok(Transaction::new(store.try_borrow()?))
+        if let Some(store) = store.upgrade() {
+            let store_ref = store.try_borrow()?;
+            let store_ref: Ref<'a, Store> = unsafe { std::mem::transmute(store_ref) };
+            Ok(Transaction::new(store_ref))
+        } else {
+            Err(TransactionAcqError::DocumentDropped)
+        }
     }
 
-    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError> {
+    fn try_transact_mut<'a>(&'a self) -> Result<TransactionMut<'a>, TransactionAcqError> {
         let store = self.store.as_ref().unwrap();
-        Ok(TransactionMut::new(store.try_borrow_mut()?))
+        if let Some(store) = store.upgrade() {
+            let store_ref = store.try_borrow_mut()?;
+            let store_ref: RefMut<'a, Store> = unsafe { std::mem::transmute(store_ref) };
+            Ok(TransactionMut::new(store_ref))
+        } else {
+            Err(TransactionAcqError::DocumentDropped)
```

Comment:
Since branches now have only weak references to store, if the corresponding doc was dropped, it may turn out that we're not longer able to create transaction (which requires store).